meldekortdomene/src/main/kotlin/no/nav/aap/varsel/VarselService.kt

```diff
@@ -0,0 +1,168 @@
+import no.nav.aap.kelvin.KelvinSakService
+import no.nav.aap.komponenter.repository.RepositoryProvider
+import no.nav.aap.lookup.gateway.GatewayProvider
+import no.nav.aap.meldeperiode.Meldeperiode
+import no.nav.aap.sak.FagsakReferanse
+import no.nav.aap.sak.Fagsaknummer
+import no.nav.aap.sak.FagsystemNavn
+import no.nav.aap.utfylling.Utfylling
+import org.slf4j.LoggerFactory
+import java.time.Clock
+import java.time.Instant
+import no.nav.aap.komponenter.config.requiredConfigForKey
+import no.nav.aap.utfylling.UtfyllingRepository
+import java.time.LocalDate
+import java.time.LocalTime
+import java.time.ZoneId
+import java.util.*
+
+class VarselService(
+    private val kelvinSakService: KelvinSakService,
+    private val kelvinSakRepository: KelvinSakRepository,
+    private val varselRepository: VarselRepository,
+    private val utfyllingRepository: UtfyllingRepository,
+    private val varselGateway: VarselGateway,
+    private val clock: Clock
+) {
+    constructor(repositoryProvider: RepositoryProvider, gatewayProvider: GatewayProvider, clock: Clock) : this(
+        kelvinSakService = KelvinSakService(repositoryProvider, gatewayProvider, clock),
+        kelvinSakRepository = repositoryProvider.provide(),
+        varselRepository = repositoryProvider.provide(),
+        utfyllingRepository = repositoryProvider.provide(),
+        varselGateway = gatewayProvider.provide(),
+        clock = clock
+    )
+
+    private val log = LoggerFactory.getLogger(javaClass)
+
+    fun planleggFremtidigeVarsler(saksnummer: Fagsaknummer) {
+        val meldeplikt = kelvinSakRepository.hentMeldeplikt(saksnummer)
+
+        val meldeperioder = kelvinSakService.hentMeldeperioder(FagsakReferanse(FagsystemNavn.KELVIN, saksnummer))
+            .filter { meldeplikt.contains(it.meldevindu) }
+
+        if (meldeperioder.size != meldeplikt.size) {
+            log.warn("Meldevindu i meldeperioder samsvarer ikke med meldeplikt-perioder fra Kelvin")
+        }
+
+        varselRepository.slettPlanlagteVarsler(saksnummer, TypeVarselOm.MELDEPLIKTPERIODE)
+        inaktiverVarslerForFjernetMeldeplikt(saksnummer, meldeperioder)
+        lagreFremtidigeVarsler(saksnummer, meldeperioder)
+    }
+
+    private fun inaktiverVarslerForFjernetMeldeplikt(saksnummer: Fagsaknummer, meldeperioder: List<Meldeperiode>) {
+        varselRepository.hentVarsler(saksnummer)
+            .filter { it.status == VarselStatus.SENDT }
+            .filterNot { varsel -> meldeperioder.map { it.meldeperioden }.contains(varsel.forPeriode) }
+            .forEach { varsel ->
+                inaktiverVarsel(varsel)
+            }
+    }
+
+    private fun lagreFremtidigeVarsler(saksnummer: Fagsaknummer, meldeperioder: List<Meldeperiode>) {
+        val avsluttedeUtfyllinger = utfyllingRepository.hentUtfyllinger(saksnummer).filter { it.erAvsluttet }
+        val fremtidigeMeldeperioder =
+            meldeperioder.filter { it.meldevindu.fom.isAfter(LocalDate.now(clock)) }
```

Comment:
Sanity check: isAfter er strengt større?

Reply:
Good catch! La til en endring nå slik at det planlegges/lagres varsel for en periode der vi står i meldevinduet _nå_, med mindre det ikke allerede er sendt varsel. I praksis sendes varselet kl 9:00 (+tiden jobben bruker på å plukke opp varsler som skal sendes) første meldedag, så vi kan havne i situasjoner det varselet er planlagt (før 9) eller sendt (eter 9), på første meldedag.